File: examples/go/patterns/parallel-subflows/flows.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package parallelsubflows

import (
	"context"
	"errors"
	"fmt"
	"hash/fnv"
	"time"

	"github.com/superdurable/dex/sdk-go/dex"
)

const (
	DefaultConcurrency  = 10
	MaxBufferedRequests = 100
)

var (
	RequestChannel     = dex.DefineChannel[string]("RequestChannel")
	Stopped            = dex.DefineAttribute[bool]("Stopped")
	CurrSubFlowNum     = dex.DefineAttribute[int]("CurrSubFlowNum")
	SubFlowCompletedCh = dex.DefineChannel[bool]("SubFlowCompletedCh")
	AllDoneCh          = dex.DefineChannel[bool]("AllDoneCh")
)

type ParentInput struct {
	Requests    []string `json:"requests"`
	Concurrency int      `json:"concurrency"`
}

type SubmitRequestInput struct {
	Request   string   `json:"request"`
	ParentIDs []string `json:"parentIds"`
}

type ExampleSubFlow struct{ dex.FlowDefaults }

func NewExampleSubFlow() *ExampleSubFlow { return &ExampleSubFlow{} }

func (*ExampleSubFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{dex.DefineStartStep(doWorkStep{})}
}

func (*ExampleSubFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{}
}

type doWorkStep struct {
	dex.StepDefaultsNoWaitFor[string]
}

func (doWorkStep) GetStepType() string { return "DoWorkStep" }

func (doWorkStep) Execute(_ dex.Context, request string) (*dex.StepDecision, error) {
	time.Sleep(time.Duration(50+len(request)%10*50) * time.Millisecond)
	return dex.GracefulComplete(request), nil
}

type BasicParentFlow struct {
	dex.FlowDefaults
	exampleFlow *ExampleSubFlow
}

func NewBasicParentFlow(exampleFlow *ExampleSubFlow) *BasicParentFlow {
	return &BasicParentFlow{exampleFlow: exampleFlow}
}

func (flow *BasicParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{dex.DefineStartStep(subFlowsStep{exampleFlow: flow.exampleFlow})}
}

func (*BasicParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{}
}

type subFlowsStep struct {
	dex.StepDefaults
	exampleFlow *ExampleSubFlow
}

func (subFlowsStep) GetStepType() string { return "SubFlowsStep" }

func (step subFlowsStep) WaitFor(_ dex.Context, requests []string) (*dex.Wait, error) {
	conditions := make([]dex.Condition, 0, len(requests))
	for _, request := range requests {
		conditions = append(conditions, dex.SubFlow(step.exampleFlow, request))
	}
	return dex.AllOf(conditions...), nil
}

func (subFlowsStep) Execute(_ dex.Context, _ []string) (*dex.StepDecision, error) {
	return dex.GracefulComplete(nil), nil
}

type WaitForHalfParentFlow struct {
	dex.FlowDefaults
	getClient   func() *dex.Client
	exampleFlow *ExampleSubFlow
}

func NewWaitForHalfParentFlow(
	getClient func() *dex.Client,
	exampleFlow *ExampleSubFlow,
) *WaitForHalfParentFlow {
	return &WaitForHalfParentFlow{getClient: getClient, exampleFlow: exampleFlow}
}

func (flow *WaitForHalfParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(waitForHalfInitStep{}),
		dex.DefineStep(subFlowStep{getClient: flow.getClient, exampleFlow: flow.exampleFlow}),
		dex.DefineStep(waitSubFlowsStep{}),
	}
}

func (*WaitForHalfParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{Channels: []dex.ChannelDef{SubFlowCompletedCh, AllDoneCh}}
}

type waitForHalfInitStep struct {
	dex.StepDefaultsNoWaitFor[[]string]
}

func (waitForHalfInitStep) GetStepType() string { return "InitStep" }

func (waitForHalfInitStep) Execute(_ dex.Context, requests []string) (*dex.StepDecision, error) {
	if len(requests) == 0 {
		return dex.GracefulComplete(nil), nil
	}
	movements := make([]dex.StepMovement, 0, len(requests)+1)
	movements = append(movements, dex.MovementOf(waitSubFlowsStep{}, len(requests)))
	for _, request := range requests {
		movements = append(movements, dex.MovementOf(subFlowStep{}, request))
	}
	return dex.GoToMany(movements...), nil
}

type subFlowStep struct {
	dex.StepDefaults
	getClient   func() *dex.Client
	exampleFlow *ExampleSubFlow
}

func (subFlowStep) GetStepType() string { return "SubFlowStep" }

func (step subFlowStep) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.AnyOf(dex.SubFlow(step.exampleFlow, request), AllDoneCh.ForOne()), nil
}

func (step subFlowStep) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	result, err := dex.SubFlowResult(ctx)
	if err != nil {
		return nil, err
	}
	if result.Status != dex.FlowRunning {
		if err := SubFlowCompletedCh.Publish(ctx, true); err != nil {
			return nil, err
		}
		return dex.GracefulComplete(nil), nil
	}
	client := step.getClient()
	if client == nil {
		return nil, fmt.Errorf("dex client is not available")
	}
	flowID, err := dex.SubFlowID(ctx)
	if err != nil {
		return nil, err
	}
	if err := client.StopFlow(context.Background(), flowID, dex.StopOptions{
		Type: dex.CancelFlow, Reason: "enough SubFlows completed",
	}); err != nil {
		return nil, err
	}
	return dex.GracefulComplete(nil), nil
}

type waitSubFlowsStep struct{ dex.StepDefaults }

func (waitSubFlowsStep) GetStepType() string { return "WaitSubFlowsStep" }

func (waitSubFlowsStep) WaitFor(_ dex.Context, total int) (*dex.Wait, error) {
	return dex.Until(SubFlowCompletedCh.ForN((total + 1) / 2)), nil
}

func (waitSubFlowsStep) Execute(ctx dex.Context, total int) (*dex.StepDecision, error) {
	remaining := total - (total+1)/2
	for index := 0; index < remaining; index++ {
		if err := AllDoneCh.Publish(ctx, true); err != nil {
			return nil, err
		}
	}
	return dex.GracefulComplete(nil), nil
}

type AdvancedLongLiveParentFlow struct {
	dex.FlowDefaults
	exampleFlow *ExampleSubFlow
}

func NewAdvancedLongLiveParentFlow(exampleFlow *ExampleSubFlow) *AdvancedLongLiveParentFlow {
	return &AdvancedLongLiveParentFlow{exampleFlow: exampleFlow}
}

func (flow *AdvancedLongLiveParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(longLiveInitStep{}),
		dex.DefineStep(longLiveHandleRequestStep{}),
		dex.DefineStep(longLiveHandleSubFlowStep{exampleFlow: flow.exampleFlow}),
	}
}

func (*AdvancedLongLiveParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{Attributes: []dex.AttributeDef{Stopped}, Channels: []dex.ChannelDef{RequestChannel}}
}

func (*AdvancedLongLiveParentFlow) SendRequest(ctx dex.Context, request string) (*dex.RPCResult[bool], error) {
	if RequestChannel.Size(ctx) >= MaxBufferedRequests {
		return &dex.RPCResult[bool]{Output: false}, nil
	}
	if err := RequestChannel.Publish(ctx, request); err != nil {
		return nil, err
	}
	return &dex.RPCResult[bool]{Output: true}, nil
}

func (*AdvancedLongLiveParentFlow) Stop(ctx dex.Context, _ dex.None) (*dex.RPCResult[dex.None], error) {
	if err := Stopped.Set(ctx, true); err != nil {
		return nil, err
	}
	return &dex.RPCResult[dex.None]{}, nil
}

type longLiveInitStep struct {
	dex.StepDefaultsNoWaitFor[ParentInput]
}

func (longLiveInitStep) GetStepType() string { return "InitStep" }

func (longLiveInitStep) Execute(ctx dex.Context, input ParentInput) (*dex.StepDecision, error) {
	for _, request := range input.Requests {
		if err := RequestChannel.Publish(ctx, request); err != nil {
			return nil, err
		}
	}
	if err := Stopped.Set(ctx, false); err != nil {
		return nil, err
	}
	return dex.GoToMany(longLiveWorkerMovements(concurrency(input.Concurrency))...), nil
}

type longLiveHandleRequestStep struct{ dex.StepDefaults }

func (longLiveHandleRequestStep) GetStepType() string { return "HandleRequestStep" }

func (longLiveHandleRequestStep) WaitFor(_ dex.Context, _ dex.None) (*dex.Wait, error) {
	return dex.Until(RequestChannel.ForOne()), nil
}

func (longLiveHandleRequestStep) Execute(ctx dex.Context, _ dex.None) (*dex.StepDecision, error) {
	requests, err := RequestChannel.GetConditionResults(ctx)
	if err != nil {
		return nil, err
	}
	return dex.GoTo(longLiveHandleSubFlowStep{}, requests[0]), nil
}

type longLiveHandleSubFlowStep struct {
	dex.StepDefaults
	exampleFlow *ExampleSubFlow
}

func (longLiveHandleSubFlowStep) GetStepType() string { return "HandleSubFlowStep" }

func (step longLiveHandleSubFlowStep) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.Until(dex.SubFlow(step.exampleFlow, request)), nil
}

func (longLiveHandleSubFlowStep) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	stopped, err := Stopped.Get(ctx)
	if err != nil {
		return nil, err
	}
	if stopped {
		return dex.GracefulComplete(nil), nil
	}
	return dex.GoTo(longLiveHandleRequestStep{}, nil), nil
}

type AdvancedShortLiveParentFlow struct {
	dex.FlowDefaults
	exampleFlow *ExampleSubFlow
}

func NewAdvancedShortLiveParentFlow(exampleFlow *ExampleSubFlow) *AdvancedShortLiveParentFlow {
	return &AdvancedShortLiveParentFlow{exampleFlow: exampleFlow}
}

func (flow *AdvancedShortLiveParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(shortLiveInitStep{}),
		dex.DefineStep(shortLiveHandleRequestStep{}),
		dex.DefineStep(shortLiveHandleSubFlowStep{exampleFlow: flow.exampleFlow}),
	}
}

func (*AdvancedShortLiveParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{Attributes: []dex.AttributeDef{CurrSubFlowNum}, Channels: []dex.ChannelDef{RequestChannel}}
}

func (*AdvancedShortLiveParentFlow) SendRequest(ctx dex.Context, request string) (*dex.RPCResult[bool], error) {
	if RequestChannel.Size(ctx) >= MaxBufferedRequests {
		return &dex.RPCResult[bool]{Output: false}, nil
	}
	if err := RequestChannel.Publish(ctx, request); err != nil {
		return nil, err
	}
	return &dex.RPCResult[bool]{Output: true}, nil
}

type shortLiveInitStep struct {
	dex.StepDefaultsNoWaitFor[ParentInput]
}

func (shortLiveInitStep) GetStepType() string { return "InitStep" }

func (shortLiveInitStep) Execute(ctx dex.Context, input ParentInput) (*dex.StepDecision, error) {
	for _, request := range input.Requests {
		if err := RequestChannel.Publish(ctx, request); err != nil {
			return nil, err
		}
	}
	if err := CurrSubFlowNum.Set(ctx, 0); err != nil {
		return nil, err
	}
	return dex.GoToMany(shortLiveWorkerMovements(concurrency(input.Concurrency))...), nil
}

type shortLiveHandleRequestStep struct{ dex.StepDefaults }

func (shortLiveHandleRequestStep) GetStepType() string { return "HandleRequestStep" }

func (shortLiveHandleRequestStep) GetStepOptions() *dex.StepOptions {
	return &dex.StepOptions{ExecuteLockAttributes: []dex.AttributeLock{dex.LockAttribute(CurrSubFlowNum)}}
}

func (shortLiveHandleRequestStep) WaitFor(_ dex.Context, _ dex.None) (*dex.Wait, error) {
	return dex.Until(RequestChannel.ForOne()), nil
}

func (shortLiveHandleRequestStep) Execute(ctx dex.Context, _ dex.None) (*dex.StepDecision, error) {
	requests, err := RequestChannel.GetConditionResults(ctx)
	if err != nil {
		return nil, err
	}
	current, err := CurrSubFlowNum.Get(ctx)
	if err != nil {
		return nil, err
	}
	if err := CurrSubFlowNum.Set(ctx, current+1); err != nil {
		return nil, err
	}
	return dex.GoTo(shortLiveHandleSubFlowStep{}, requests[0]), nil
}

type shortLiveHandleSubFlowStep struct {
	dex.StepDefaults
	exampleFlow *ExampleSubFlow
}

func (shortLiveHandleSubFlowStep) GetStepType() string { return "HandleSubFlowStep" }

func (shortLiveHandleSubFlowStep) GetStepOptions() *dex.StepOptions {
	return &dex.StepOptions{ExecuteLockAttributes: []dex.AttributeLock{dex.LockAttribute(CurrSubFlowNum)}}
}

func (step shortLiveHandleSubFlowStep) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.Until(dex.SubFlow(step.exampleFlow, request)), nil
}

func (shortLiveHandleSubFlowStep) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	current, err := CurrSubFlowNum.Get(ctx)
	if err != nil {
		return nil, err
	}
	current--
	if err := CurrSubFlowNum.Set(ctx, current); err != nil {
		return nil, err
	}
	if current == 0 {
		return dex.ForceCompleteIfChannelsEmpty(
			nil,
			[]dex.ChannelDef{RequestChannel},
			dex.MovementOf(shortLiveHandleRequestStep{}, nil),
		), nil
	}
	return dex.GoTo(shortLiveHandleRequestStep{}, nil), nil
}

func longLiveWorkerMovements(count int) []dex.StepMovement {
	movements := make([]dex.StepMovement, 0, count)
	for index := 0; index < count; index++ {
		movements = append(movements, dex.MovementOf(longLiveHandleRequestStep{}, nil))
	}
	return movements
}

func shortLiveWorkerMovements(count int) []dex.StepMovement {
	movements := make([]dex.StepMovement, 0, count)
	for index := 0; index < count; index++ {
		movements = append(movements, dex.MovementOf(shortLiveHandleRequestStep{}, nil))
	}
	return movements
}

func concurrency(configured int) int {
	if configured > 0 {
		return configured
	}
	return DefaultConcurrency
}

type SubmitRequestFlow struct {
	dex.FlowDefaults
	getClient  func() *dex.Client
	parentFlow *AdvancedShortLiveParentFlow
}

func NewSubmitRequestFlow(
	getClient func() *dex.Client,
	parentFlow *AdvancedShortLiveParentFlow,
) *SubmitRequestFlow {
	return &SubmitRequestFlow{getClient: getClient, parentFlow: parentFlow}
}

func (flow *SubmitRequestFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{dex.DefineStartStep(submitStep{
		getClient: flow.getClient, parentFlow: flow.parentFlow,
	})}
}

func (*SubmitRequestFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{}
}

type submitStep struct {
	dex.StepDefaultsNoWaitFor[SubmitRequestInput]
	getClient  func() *dex.Client
	parentFlow *AdvancedShortLiveParentFlow
}

func (submitStep) GetStepType() string { return "SubmitStep" }

func (step submitStep) Execute(_ dex.Context, input SubmitRequestInput) (*dex.StepDecision, error) {
	if len(input.ParentIDs) == 0 {
		return nil, fmt.Errorf("at least one parent Flow ID is required")
	}
	client := step.getClient()
	if client == nil {
		return nil, fmt.Errorf("dex client is not available")
	}
	parentID := input.ParentIDs[partition(input.Request, len(input.ParentIDs))]
	accepted, err := enqueueRequest(
		context.Background(), client, step.parentFlow, parentID, input.Request,
	)
	if err != nil {
		return nil, err
	}
	if !accepted {
		return nil, fmt.Errorf("parent %s rejected the request", parentID)
	}
	return dex.GracefulComplete(parentID), nil
}

func enqueueRequest(
	ctx context.Context,
	client *dex.Client,
	parentFlow *AdvancedShortLiveParentFlow,
	parentID string,
	request string,
) (bool, error) {
	accepted, err := invokeRequest(ctx, client, parentFlow, parentID, request)
	if err == nil {
		return accepted, nil
	}
	var inactive *dex.FlowNotActiveError
	if !errors.As(err, &inactive) {
		return false, err
	}
	_, err = client.StartFlow(ctx, parentFlow, parentID, ParentInput{
		Requests: []string{request}, Concurrency: DefaultConcurrency,
	}, dex.StartFlowOptions{IDReusePolicy: dex.IDReuseAllowIfNotRunning})
	if err == nil {
		return true, nil
	}
	var alreadyStarted *dex.FlowAlreadyStartedError
	if !errors.As(err, &alreadyStarted) {
		return false, err
	}
	return invokeRequest(ctx, client, parentFlow, parentID, request)
}

func invokeRequest(
	ctx context.Context,
	client *dex.Client,
	parentFlow *AdvancedShortLiveParentFlow,
	parentID string,
	request string,
) (bool, error) {
	var accepted bool
	err := client.InvokeRPC(
		ctx, parentID, parentFlow.SendRequest, request, &accepted, dex.InvokeOptions{},
	)
	return accepted, err
}

func partition(request string, partitions int) int {
	hash := fnv.New32a()
	_, _ = hash.Write([]byte(request))
	return int(hash.Sum32()) % partitions
}

var (
	_ dex.Flow                    = (*ExampleSubFlow)(nil)
	_ dex.Flow                    = (*BasicParentFlow)(nil)
	_ dex.Flow                    = (*WaitForHalfParentFlow)(nil)
	_ dex.Flow                    = (*AdvancedLongLiveParentFlow)(nil)
	_ dex.Flow                    = (*AdvancedShortLiveParentFlow)(nil)
	_ dex.Flow                    = (*SubmitRequestFlow)(nil)
	_ dex.RPC[string, bool]       = (*AdvancedLongLiveParentFlow)(nil).SendRequest
	_ dex.RPC[dex.None, dex.None] = (*AdvancedLongLiveParentFlow)(nil).Stop
	_ dex.RPC[string, bool]       = (*AdvancedShortLiveParentFlow)(nil).SendRequest
)

File: examples/python/dex_examples/patterns/parallel-subflows/flows.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Callable

from dex import (
    AsyncClient,
    Attribute,
    Channel,
    Context,
    Flow,
    FlowNotActiveError,
    FlowStatus,
    IdReusePolicy,
    PersistenceSchema,
    RPCResult,
    Step,
    StepDecision,
    StepList,
    StepMovement,
    StepOptions,
    StartFlowOptions,
    SubFlow,
    Wait,
    force_complete_if_channels_empty,
    go_to,
    go_to_many,
    graceful_complete,
    rpc,
)

DEFAULT_CONCURRENCY = 10
MAX_BUFFERED_REQUESTS = 100


@dataclass(frozen=True)
class ParentInput:
    requests: list[str]
    concurrency: int = DEFAULT_CONCURRENCY


@dataclass(frozen=True)
class SubmitRequestInput:
    request: str
    parent_ids: list[str]


class DoWorkStep(Step[str]):
    async def execute(self, context: Context, request: str) -> StepDecision:
        await asyncio.sleep((50 + len(request) % 10 * 50) / 1000)
        return graceful_complete(request)


class ExampleSubFlow(Flow[str]):
    def __init__(self) -> None:
        self.do_work = DoWorkStep()

    def get_steps(self) -> StepList[str]:
        return StepList.start_step(self.do_work)


class SubFlowsStep(Step[list[str]]):
    def __init__(self, example_subflow: ExampleSubFlow) -> None:
        self.example_subflow = example_subflow

    def wait_for(self, context: Context, requests: list[str]) -> Wait:
        return Wait.all_of(
            *(SubFlow.run(self.example_subflow, request) for request in requests)
        )

    def execute(self, context: Context, requests: list[str]) -> StepDecision:
        return graceful_complete()


class BasicParentFlow(Flow[list[str]]):
    def __init__(self, example_subflow: ExampleSubFlow) -> None:
        self.subflows = SubFlowsStep(example_subflow)

    def get_steps(self) -> StepList[list[str]]:
        return StepList.start_step(self.subflows)


class WaitForHalfInitStep(Step[list[str]]):
    def execute(self, context: Context, requests: list[str]) -> StepDecision:
        if not requests:
            return graceful_complete()
        return go_to_many(
            StepMovement.of(WaitSubFlowsStep, len(requests)),
            *(StepMovement.of(SubFlowStep, request) for request in requests),
        )


class SubFlowStep(Step[str]):
    def __init__(
        self,
        client_provider: Callable[[], AsyncClient],
        example_subflow: ExampleSubFlow,
        subflow_completed_ch: Channel[bool],
        all_done_ch: Channel[bool],
    ) -> None:
        self.client_provider = client_provider
        self.example_subflow = example_subflow
        self.subflow_completed_ch = subflow_completed_ch
        self.all_done_ch = all_done_ch

    def wait_for(self, context: Context, request: str) -> Wait:
        return Wait.any_of(
            SubFlow.run(self.example_subflow, request), self.all_done_ch.for_one()
        )

    async def execute(  # type: ignore[override]
        self, context: Context, request: str
    ) -> StepDecision:
        result = SubFlow.get_condition_results(context)
        if result.status is not FlowStatus.RUNNING:
            self.subflow_completed_ch.publish(context, True)
            return graceful_complete()
        await self.client_provider().stop_flow(SubFlow.get_flow_id(context))
        return graceful_complete()


class WaitSubFlowsStep(Step[int]):
    def __init__(
        self, subflow_completed_ch: Channel[bool], all_done_ch: Channel[bool]
    ) -> None:
        self.subflow_completed_ch = subflow_completed_ch
        self.all_done_ch = all_done_ch

    def wait_for(self, context: Context, total: int) -> Wait:
        return Wait.until(self.subflow_completed_ch.for_n((total + 1) // 2))

    def execute(self, context: Context, total: int) -> StepDecision:
        for _ in range(total - (total + 1) // 2):
            self.all_done_ch.publish(context, True)
        return graceful_complete()


class WaitForHalfParentFlow(Flow[list[str]]):
    subflow_completed_ch = Channel("SubFlowCompletedCh", bool)
    all_done_ch = Channel("AllDoneCh", bool)

    def __init__(
        self,
        client_provider: Callable[[], AsyncClient],
        example_subflow: ExampleSubFlow,
    ) -> None:
        self.init = WaitForHalfInitStep()
        self.subflow = SubFlowStep(
            client_provider, example_subflow, self.subflow_completed_ch, self.all_done_ch
        )
        self.wait_subflows = WaitSubFlowsStep(
            self.subflow_completed_ch, self.all_done_ch
        )

    def get_steps(self) -> StepList[list[str]]:
        return StepList.start_step(self.init).other_steps(
            self.subflow, self.wait_subflows
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.subflow_completed_ch, self.all_done_ch)


class LongLiveInitStep(Step[ParentInput]):
    def __init__(
        self,
        request_channel: Channel[str],
        stopped: Attribute[bool],
    ) -> None:
        self.request_channel = request_channel
        self.stopped = stopped

    def get_step_type(self) -> str:
        return "InitStep"

    def execute(self, context: Context, input: ParentInput) -> StepDecision:
        for request in input.requests:
            self.request_channel.publish(context, request)
        self.stopped.set(context, False)
        concurrency = input.concurrency if input.concurrency > 0 else DEFAULT_CONCURRENCY
        return go_to_many(
            *(StepMovement.of(LongLiveHandleRequestStep, None) for _ in range(concurrency))
        )


class LongLiveHandleRequestStep(Step[None]):
    def __init__(self, request_channel: Channel[str]) -> None:
        self.request_channel = request_channel

    def get_step_type(self) -> str:
        return "HandleRequestStep"

    def wait_for(self, context: Context, input: None) -> Wait:
        return Wait.until(self.request_channel.for_one())

    def execute(self, context: Context, input: None) -> StepDecision:
        return go_to(LongLiveHandleSubFlowStep, self.request_channel.results(context)[0])


class LongLiveHandleSubFlowStep(Step[str]):
    def __init__(
        self,
        example_subflow: ExampleSubFlow,
        stopped: Attribute[bool],
    ) -> None:
        self.example_subflow = example_subflow
        self.stopped = stopped

    def get_step_type(self) -> str:
        return "HandleSubFlowStep"

    def wait_for(self, context: Context, request: str) -> Wait:
        return Wait.until(SubFlow.run(self.example_subflow, request))

    def execute(self, context: Context, request: str) -> StepDecision:
        if self.stopped.get(context):
            return graceful_complete()
        return go_to(LongLiveHandleRequestStep, None)


class AdvancedLongLiveParentFlow(Flow[ParentInput]):
    request_channel = Channel("RequestChannel", str)
    stopped = Attribute("Stopped", bool)

    def __init__(self, example_subflow: ExampleSubFlow) -> None:
        self.init = LongLiveInitStep(self.request_channel, self.stopped)
        self.handle_request = LongLiveHandleRequestStep(self.request_channel)
        self.handle_subflow = LongLiveHandleSubFlowStep(example_subflow, self.stopped)

    def get_steps(self) -> StepList[ParentInput]:
        return StepList.start_step(self.init).other_steps(
            self.handle_request, self.handle_subflow
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.request_channel, self.stopped)

    @rpc
    def send_request(self, context: Context, request: str) -> RPCResult[bool]:
        if self.request_channel.size(context) >= MAX_BUFFERED_REQUESTS:
            return RPCResult(False)
        self.request_channel.publish(context, request)
        return RPCResult(True)

    @rpc
    def stop(self, context: Context) -> None:
        self.stopped.set(context, True)


class ShortLiveInitStep(Step[ParentInput]):
    def __init__(
        self,
        request_channel: Channel[str],
        curr_subflow_num: Attribute[int],
    ) -> None:
        self.request_channel = request_channel
        self.curr_subflow_num = curr_subflow_num

    def get_step_type(self) -> str:
        return "InitStep"

    def execute(self, context: Context, input: ParentInput) -> StepDecision:
        for request in input.requests:
            self.request_channel.publish(context, request)
        self.curr_subflow_num.set(context, 0)
        concurrency = input.concurrency if input.concurrency > 0 else DEFAULT_CONCURRENCY
        return go_to_many(
            *(StepMovement.of(ShortLiveHandleRequestStep, None) for _ in range(concurrency))
        )


class ShortLiveHandleRequestStep(Step[None]):
    def __init__(
        self,
        request_channel: Channel[str],
        curr_subflow_num: Attribute[int],
    ) -> None:
        self.request_channel = request_channel
        self.curr_subflow_num = curr_subflow_num

    def get_step_type(self) -> str:
        return "HandleRequestStep"

    def get_step_options(self) -> StepOptions:
        return StepOptions(execute_lock_attributes=(self.curr_subflow_num.lock(),))

    def wait_for(self, context: Context, input: None) -> Wait:
        return Wait.until(self.request_channel.for_one())

    def execute(self, context: Context, input: None) -> StepDecision:
        request = self.request_channel.results(context)[0]
        self.curr_subflow_num.set(context, (self.curr_subflow_num.get(context) or 0) + 1)
        return go_to(ShortLiveHandleSubFlowStep, request)


class ShortLiveHandleSubFlowStep(Step[str]):
    def __init__(
        self,
        example_subflow: ExampleSubFlow,
        request_channel: Channel[str],
        curr_subflow_num: Attribute[int],
    ) -> None:
        self.example_subflow = example_subflow
        self.request_channel = request_channel
        self.curr_subflow_num = curr_subflow_num

    def get_step_type(self) -> str:
        return "HandleSubFlowStep"

    def get_step_options(self) -> StepOptions:
        return StepOptions(execute_lock_attributes=(self.curr_subflow_num.lock(),))

    def wait_for(self, context: Context, request: str) -> Wait:
        return Wait.until(SubFlow.run(self.example_subflow, request))

    def execute(self, context: Context, request: str) -> StepDecision:
        current = (self.curr_subflow_num.get(context) or 0) - 1
        self.curr_subflow_num.set(context, current)
        if current == 0:
            return force_complete_if_channels_empty(
                None,
                StepMovement.of(ShortLiveHandleRequestStep, None),
                self.request_channel,
            )
        return go_to(ShortLiveHandleRequestStep, None)


class AdvancedShortLiveParentFlow(Flow[ParentInput]):
    request_channel = Channel("RequestChannel", str)
    curr_subflow_num = Attribute("CurrSubFlowNum", int)

    def __init__(self, example_subflow: ExampleSubFlow) -> None:
        self.init = ShortLiveInitStep(self.request_channel, self.curr_subflow_num)
        self.handle_request = ShortLiveHandleRequestStep(
            self.request_channel, self.curr_subflow_num
        )
        self.handle_subflow = ShortLiveHandleSubFlowStep(
            example_subflow, self.request_channel, self.curr_subflow_num
        )

    def get_steps(self) -> StepList[ParentInput]:
        return StepList.start_step(self.init).other_steps(
            self.handle_request, self.handle_subflow
        )

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.request_channel, self.curr_subflow_num)

    @rpc
    def send_request(self, context: Context, request: str) -> RPCResult[bool]:
        if self.request_channel.size(context) >= MAX_BUFFERED_REQUESTS:
            return RPCResult(False)
        self.request_channel.publish(context, request)
        return RPCResult(True)


class SubmitStep(Step[SubmitRequestInput]):
    def __init__(
        self,
        client_provider: Callable[[], AsyncClient],
        parent_flow: AdvancedShortLiveParentFlow,
    ) -> None:
        self.client_provider = client_provider
        self.parent_flow = parent_flow

    async def execute(  # type: ignore[override]
        self, context: Context, input: SubmitRequestInput
    ) -> StepDecision:
        if not input.parent_ids:
            raise ValueError("at least one parent Flow ID is required")
        parent_id = input.parent_ids[partition(input.request, len(input.parent_ids))]
        accepted = await enqueue_request(
            self.client_provider(), self.parent_flow, parent_id, input.request
        )
        if not accepted:
            raise RuntimeError(f"parent {parent_id} rejected the request")
        return graceful_complete(parent_id)


async def enqueue_request(
    client: AsyncClient,
    parent_flow: AdvancedShortLiveParentFlow,
    parent_id: str,
    request: str,
) -> bool:
    try:
        return await client.invoke_rpc(parent_flow.send_request, parent_id, request)
    except FlowNotActiveError:
        await client.start_flow(
            parent_flow,
            parent_id,
            ParentInput([request], DEFAULT_CONCURRENCY),
            StartFlowOptions(id_reuse_policy=IdReusePolicy.ALLOW_IF_NOT_RUNNING),
        )
        return True


def partition(request: str, partitions: int) -> int:
    hash_value = 2_166_136_261
    for byte in request.encode():
        hash_value ^= byte
        hash_value = hash_value * 16_777_619 & 0xFFFFFFFF
    return hash_value % partitions


class SubmitRequestFlow(Flow[SubmitRequestInput]):
    def __init__(
        self,
        client_provider: Callable[[], AsyncClient],
        parent_flow: AdvancedShortLiveParentFlow,
    ) -> None:
        self.submit = SubmitStep(client_provider, parent_flow)

    def get_steps(self) -> StepList[SubmitRequestInput]:
        return StepList.start_step(self.submit)
